### PR TITLE
fix(babel): fix unused platform code removal

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

Metro applies this plugin too but it does it after the imports have been transformed which breaks the plugin. Here, we'll apply it before the commonjs transform, in production, to ensure `Platform.OS` is replaced with a string literal and `__DEV__` is converted to a boolean.
Applying early also means that web can be transformed before the `react-native-web` transform mutates the import.

# Test Plan

- Added new tests for the plugin.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by brentvatne — APPROVED** [review]

File: packages/babel-preset-expo/src/__tests__/platform-shaking.test.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    ...DEFAULT_OPTS,
+    caller: getCaller({ name: 'metro', engine: 'hermes', platform: 'web' }),
+  };
+
+  expect(
+    babel.transform(
+      `Platform.select({ ios: () => console.log('ios'), web: () => console.log('web'), android: () => console.log('android'), })`,
+      options
+    )!.code
+  ).toEqual(`(function(){return console.log('web');});`);
```
Comment (ide) [inline packages/babel-preset-expo/src/__tests__/platform-shaking.test.ts:44]: Any reason to keep this code when the code right below tests a superset of Platform.select functionality?

**Review by ide — APPROVED** [review]

**Review by ide — APPROVED** [review]

File: packages/babel-preset-expo/src/index.ts [hunk 2]
```diff
@@ -98,6 +98,21 @@ function babelPresetExpo(api: ConfigAPI, options: BabelPresetExpoOptions = {}):
     extraPlugins.push(require('@babel/plugin-transform-parameters'));
   }
 
+  if (!isDev && hasModule('metro-transform-plugins')) {
+    // Metro applies this plugin too but it does it after the imports have been transformed which breaks
+    // the plugin. Here, we'll apply it before the commonjs transform, in production, to ensure `Platform.OS`
+    // is replaced with a string literal and `__DEV__` is converted to a boolean.
+    // Applying early also means that web can be transformed before the `react-native-web` transform mutates the import.
+    extraPlugins.push([
+      require('metro-transform-plugins/src/inline-plugin.js'),
```
Comment (ide) [inline packages/babel-preset-expo/src/index.ts:107]: Seems better to use the package's interface and load this with `require('metro-transform-plugins').inlinePlugin()`

File: packages/babel-preset-expo/src/index.ts [hunk 3]
```diff
@@ -98,6 +98,21 @@ function babelPresetExpo(api: ConfigAPI, options: BabelPresetExpoOptions = {}):
     extraPlugins.push(require('@babel/plugin-transform-parameters'));
   }
 
+  if (!isDev && hasModule('metro-transform-plugins')) {
```
Comment (ide) [inline packages/babel-preset-expo/src/index.ts:101]: Why should this not run in development, too?

Comment (EvanBacon) [reply packages/babel-preset-expo/src/index.ts:101]: We don't do any minification in development to keep things fast, might change in the future but right now it doesn't provide much benefit.

File: packages/babel-preset-expo/src/index.ts (hunk 2 above)
Comment (EvanBacon) [reply packages/babel-preset-expo/src/index.ts:107]: Pulls in more JavaScript that we aren't using, this import also recently changed from being in fbjs or something.

**Pull request merged**